python/pyspark/ml/functions.py

```diff
@@ -106,6 +117,474 @@ def array_to_vector(col: Column) -> Column:
+    # +--------------------+------------------+----+----+
+    # |                  t1|                t2|sum1|sum2|
+    # +--------------------+------------------+----+----+
+    # |[0.0, 1.0, 2.0, 3.0]|   [0.0, 1.0, 2.0]| 6.0| 3.0|
+    # |[4.0, 5.0, 6.0, 7.0]|   [4.0, 5.0, 6.0]|22.0|15.0|
+    # |[8.0, 9.0, 10.0, ...|  [8.0, 9.0, 10.0]|38.0|27.0|
+    # |[12.0, 13.0, 14.0...|[12.0, 13.0, 14.0]|54.0|39.0|
+    # |[16.0, 17.0, 18.0...|[16.0, 17.0, 18.0]|70.0|51.0|
+    # +--------------------+------------------+----+----+
+    ```
+
+    Parameters
+    ----------
+    predict_batch_fn : Callable[[],
+        Callable[..., np.ndarray | Mapping[str, np.ndarray] | List[Mapping[str, np.dtype]] ]
+        Function which is responsible for loading a model and returning a `predict` function which
+        takes one or more numpy arrays as input and returns either a numpy array (for a single
+        output), a dictionary of named numpy arrays (for multiple outputs), or a row-oriented list
+        of dictionaries (for multiple outputs).
+    return_type : :class:`pspark.sql.types.DataType` or str.
+        Spark SQL datatype for the expected output.
+    batch_size : int
+        Batch size to use for inference, note that this is typically a limitation of the model
+        and/or the hardware resources and is usually smaller than the Spark partition size.
+    input_tensor_shapes: list[list[int] | None] | Mapping[int, list[int]] | None
+        Optional input tensor shapes for models with tensor inputs.  This can be a list of shapes,
+        where each shape is a list of integers or None (for scalar inputs).  Alternatively, this
+        can be represented by a "sparse" dictionary, where the keys are the integer indices of the
+        inputs, and the values are the shapes.  Each tensor input value in the Spark DataFrame must
+        be represented as a single column containing a flattened 1-D array.  The provided
+        input_tensor_shapes will be used to reshape the flattened array into expected tensor shape.
+        For the list form, the order of the tensor shapes must match the order of the selected
+        DataFrame columns.  The batch dimension (typically -1 or None in the first dimension) should
+        not be included, since it will be determined by the batch_size argument.  Tabular datasets
+        with scalar-valued columns should not provide this argument.
+
+    Returns
+    -------
+    A pandas_udf for predicting a batch.
+    """
+    # generate a new uuid each time this is invoked on the driver to invalidate executor-side cache.
+    model_uuid = uuid.uuid4()
+
+    def predict(data: Iterator[Union[pd.Series, pd.DataFrame]]) -> Iterator[pd.DataFrame]:
+        from pyspark.ml.model_cache import ModelCache
+
+        # get predict function (from cache or from running user-provided predict_batch_fn)
+        predict_fn = ModelCache.get(model_uuid)
```

Comment:
You don't remove the model from cache when spark job completes.
This causes issue if user set `spark.python.worker.reuse` to be `true`, in this case multiple spark jobs might reuse python task process and the ModelCache might grow bigger and bigger but we cannot release model in cache.

In the opposite side, if user set `spark.python.worker.reuse` to be `false`, then the `ModelCache` becomes meaningless, in this case, each spark task it will create a new python process to run UDF routine, then in every task it need to load the model again because in new python process the ModelCache dict is always empty.

Reply:
Yes, this is expected.  The code is modeled after MLFlow's SparkModelCache (per the [prior comments](https://github.com/apache/spark/pull/37734#discussion_r960529530)), which has the same behavior/issues, AFAICT.

Reply:
We can start with it, but let's add a TODO for follow-up improvements.